### ISSUE TITLE
Slow docker save/load in v3.3.1

### ISSUE THREAD

**Issue by ykrevnyi** [concourse/concourse#1364]
# Bug Report
In the latest Concourse v3.3.1 the `docker` step with `save: true` or `load_base` parameter is pretty slow.
It seems to have problems while saving (or loading) images to/from cache.

Example:
- clone repo
- pull base image (base image size ~ 200mb)
```
Pulling <image>
...
Successfully pulled <image>
```
And it stops for 3-5 min after message `Successfully pulled <image>`.

Same thing happens when we are trying to use `load_base:` parameter.

The following can also be handy:

* Concourse version: v3.3.1
* Deployment type (BOSH/Docker/binary): binary
* Infrastructure/IaaS: aws
* Browser (if applicable): chrome
* Did this used to work? yup

**Comment by andrewedstrom** [comment]
@ykrevnyi What version of concourse were you on before this, when it was working better?

**Comment by ykrevnyi** [comment]
@andrewedstrom I was on `v2.7.7`

Yesterday updated Concourse to `v3.3.2` and problem still exists.
The CI just waiting for 15-20 minutes before every `put` step without any output at all.

**Comment by ykrevnyi** [comment]
Just double checked and can confirm that the same pipeline in `v3.3.2` is actually slower then in `v2.7.7`.

Scenario had been tested 10 time in both versions.

It is interesting that the most steps finished faster in `v3.3.2` then `v2.7.7`. But then `v3.2.2` just stuck in `put: docker-image` without any output.

**Comment by ykrevnyi** [comment]
@andrewedstrom any update on this one?
Currently we stuck at `v2.7.7` because of this issue.

/cc @vito

**Comment by gberezman** [comment]
FWIW, I'm seeing the same behavior on the versions I tried (3.4.0, 3.3.4 on a docker-compose over bare metal install). Without load_base and it's associated resource, this works fine.

With load_base, after about 12 minutes, 3.4.0 timed out with the error message: Post /volumes: net/http: timeout awaiting response headers

3.3.4 just spun, but I didn't have time to let it run for more than about 10 minutes.

I wasn't able to run 2.7.7. For some reason it couldn't connect to my git repo.

I've attached [pipeline.txt](https://github.com/concourse/concourse/files/1243876/pipeline.txt), which contains a sanitized version of the pipeline.

**Comment by drahnr** [comment]
Possibly related: 
https://github.com/concourse/concourse/issues/1405

**Comment by novcn** [comment]
Glad there's an issue for this because I too am seeing this on `v3.3.2`. Let me know if there's any debugging information I can provide.

**Comment by thetooth** [comment]
I have the same issue on 3.6.0 with or without using `save: true` or any of the cache features. In fact I've never seen a resource actually cached. The pipeline I'm using:
![screencapture-10-0-1-1-8420-teams-main-pipelines-thetooth-name-1509295871538](https://user-images.githubusercontent.com/42569/32285468-173f826c-bf76-11e7-91c3-dce711dfe933.png)

The behaviour I'm seeing is each image_resource or resource of type docker-image is pulled fully every build. After the `Successfully pulled <image>` message the build will hang, sometimes only 2 minutes but as long as 15 minutes occasionally. Finally (might be unrelated) as you can see in the pipeline `build` puts a docker image resource, after pushing and pulling in the `build` step, `test` is triggered and goes out and pulls the image for the second time and a third time for `push`, again with 2-15 minute delay before any work is done(all cores idle, minimal disk/network activity as well). All images are on a private repo in our LAN.

tl;dr it's about 10x faster to perform the pipeline steps by hand at the moment :1st_place_medal:

**Comment by EugenMayer** [comment]
Just as a little gut feeling, who is using btrfs as a fs for docker-engine?

**Comment by novcn** [comment]
@EugenMayer I am using overlay and having this issue.

**Comment by ykrevnyi** [comment]
There is the same problem using `privileged` option.
Any `task` with `privileged: true` option will have that weird timeout before start (~up to 15min).
Considering fact that `docker-image` resource is privileged it could be the issue.

**Comment by thetooth** [comment]
@EugenMayer Also using overlay on top of ext4

**Comment by EugenMayer** [comment]
I am now using https://github.com/EugenMayer/docker-image-resource/tree/develop / https://hub.docker.com/r/eugenmayer/concourse-docker-image-resource/ as a replacement for docker-image-resource.

This enabled me to pull base images from any number of registries during a build and not being in the need pre pull as a resource, save and load - this is far better then anything else.

Next thing is basically deploying a sonatype nexus docker proxy locally to the concourse stack so pushing / pulling does not even has a transport at all - its fast enough right now, but its going to be very fast then. This nexus sonatype will then also be the maven proxy / local "global" cache, so maven projcts with huge dependencies have proper build times - very similar issue to docker save/load - its just unpractical to pull 2GB of maven dependcies for one task and doing that once again the next job.

**Comment by Luc45** [comment]
Is it possible that this is a real performance bug that got ignored because of stale and persists to this day?
